core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -628,6 +621,13 @@ private[spark] class ExecutorAllocationManager(
             allocationManager.onExecutorIdle(executorId)
           }
         }
+
+        // If the task failed, we expect it to be resubmitted later.
+        if (taskEnd.reason != Success) {
+          stageIdToTaskIndices.get(stageId).foreach { taskIndices =>
+            taskIndices.remove(taskIndex)
```

Comment:
Andrew noted before:
> If anything, I would think that we should remove [this line](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala#L606). If this task fails, then the next attempt would go to the else case of stageIdNumTasks.getOrElse(stageId, -1), which not technically correct. It's safe to remove it because we remove it in stageCompleted anyway.

I agree with this option, so I delete the Line 602-610. After delete these lines, we needn't call ``` allocationManager.onSchedulerBacklogged()``` any more, because ```allocationManager.onSchedulerQueueEmpty()``` only called when all stages finish.